### PR TITLE
[NPU]dsv3.2 cp for npu

### PR DESCRIPTION
**Pull request by liupeng374**
## Motivation

Now, sglang already has the --enable-nsa-prefill-context-parallel option proposed  to support CP parallelism, [(1/n)support context parallel with deepseekv3.2-DSA ](https://github.com/sglang-npu/sglang_npu/commit/76504b8b078a7bf0a6a527632d697c1b7734feba), but NPU not supports DSV 3.2 context parallelism, This PR is intended to solve this problem.

## Modifications

This feature currently reuses the TP domain for CP, and employs a head-tail combination method to achieve CP load balancing. For details, please refer to the aforementioned PR. 

There are still some limitations at present.

1. --moe-dense-tp 1
2. --moe-a2a-backend deepep
3. only support batch size 1

## Accuracy Tests

image

## Benchmarking and Profiling

image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/distributed/parallel_state.py [hunk 1]
```diff
@@ -757,6 +757,9 @@ def cp_all_gather_into_tensor_async(
         eliminating the CPU-side launch-kernel blocking issue caused by synchronization problems.
         The specific implementation uses the interface provided by pynccl to remove the synchronization logic of events.
         """
+        if _is_npu:
+            self._all_gather_into_tensor(output, input)
+            return
```
Comment (iforgetmyname) [inline python/sglang/srt/distributed/parallel_state.py]: remove this if

File: python/sglang/srt/distributed/parallel_state.py
Comment (iforgetmyname) [inline python/sglang/srt/distributed/parallel_state.py]:
```
SUPPORT_ASYNC_ALL_GATHER = XXX
if SUPPORT_ASYNC_ALL_GATHER:
    xxxx
else:
    logger.warn
    self.all_gather_into_tensor()
```

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 2]
```diff
@@ -15,19 +15,17 @@
         import deep_gemm
     except ImportError as e:
         deep_gemm = e
-
+elif is_npu():
+    import torch_npu
+    import custom_ops  # noqa: F401
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: remove this import and move import custom_ops to `init_npu_backend()` in npu/utils.py
try-catch import custom_ops if not available

File: python/sglang/srt/layers/attention/nsa/utils.py [hunk 3]
```diff
@@ -312,17 +309,36 @@ def prepare_input_dp_with_cp_dsa(
 
     # TODO Support multi-batch-cp-split, multi-batch-cp support has accuracy issues
     # cp_seq_index = calculate_cp_seq_idx(split_list[:], seqs_len[:])
+    kv_len_prev = prefix_sum_list[cp_rank]
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/attention/nsa/utils.py:319]: use defaultdict instead
```
dev = DefaultDict(
lambda: ("cpu", torch.int64),
{"npu": ("cuda", torch.int32)}
)

torch.tensor(x).to(device=dev[device][0], dtype=dev[device][1])
```
defaultdict with namedtuple will be better

File: python/sglang/srt/distributed/parallel_state.py
Comment (liupeng374) [reply python/sglang/srt/distributed/parallel_state.py]: done

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py (hunk 2 above)
Comment (liupeng374) [reply python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: done

File: python/sglang/srt/layers/attention/nsa/utils.py (hunk 3 above)
Comment (liupeng374) [reply python/sglang/srt/layers/attention/nsa/utils.py:319]: done

File: python/sglang/srt/distributed/parallel_state.py (hunk 1 above)
Comment (liupeng374) [reply python/sglang/srt/distributed/parallel_state.py]: done

**Review by iforgetmyname — APPROVED** [review]

**Review by iforgetmyname — APPROVED** [review]
